pkg/plugin/client/client.go

```diff
@@ -66,8 +74,13 @@ func (*clientFactory) New(aspectplugin loader.AspectPlugin, streams ioutils.Stre
 	}
 
 	res := &PluginInstance{
-		Plugin:   rawplugin.(plugin.Plugin),
-		Provider: goclient,
+		Plugin:         rawplugin.(plugin.Plugin),
+		Provider:       goclient,
+		CustomCommands: map[string]*plugin.Command{},
+	}
+
+	if customCommandExecutor, ok := rawplugin.(CustomCommandExecutor); ok {
+		res.CustomCommandExecutor = customCommandExecutor
```

Comment:
I am not sure? Me being new to Go might be showing through here but the point of this is to make sure the plugin has the method `ExecuteCustomCommand` that we can call. This is not part of `Plugin` at all. Its a method added internally on the GRPC interactions so that we can run the commands provided by the plugin on the running instance of the given plugin. I dont think we want to add it to the Plugin as then it may show up to the plugins as something they could possibly implement?

Reply:
We dont want the users to know about `ExecuteCustomCommand`

Reply:
This might be me being new to Go as well...

But both `CustomCommandExecutor` and `Plugin` are just the `rawplugin` casted, so they are references to the same object? If it's always references to the same object I think it would be nice to avoid having a different field for each object type.

Some languages support things like `CustomCommandExecutor & Plugin` to indicate that it is both. I'm not sure if Go has anything similar?

Reply:
We don't want to mix these 2 interfaces. The `Plugin` interface is the public API, it's what we enforce the plugin implementations to provide.

Because of the internal juggling we do over gRPC, the runtime `Plugin`, which is the `rawplugin` is not exactly the implementation provided by the plugin author. It's our internal representation of it, i.e. the gRPC implementation that wraps the real plugin logic implementation.

CLI core -> gRPC wrapper that satisfies the Plugin interface -> real implementation that satisfies the Plugin interface

Having said that, we could use some documentation here to avoid any future confusion. Or maybe pick names that translate better? E.g. `InternalExecutor`, or put this interface inside an `internal` package.